Fix exclude_fields_in_copy with DeferringRelatedManager object

elhussienalmasri opened the pull request:
Fixes #11715

After                             |  Before
:------------------:         |    :------------------:
![After](https://github.com/wagtail/wagtail/assets/90080237/671bfa25-445d-4680-89c1-0caf069e0a5c) | ![Before](https://github.com/wagtail/wagtail/assets/90080237/f7ac70e8-d618-456a-a8ca-e24fb1d9cb9d)

lb- commented:
@elhussienalmasri are you able to add unit tests for this change. The tests should fail before this change and pass after it. So that we don't break this again in the future.

elhussienalmasri commented:
sure @lb- , it a bit complicated because its require DeferringRelatedManager object , but I will try to find similar unit test , and  I will add unit test for this PR asap.

elhussienalmasri commented:
I added @lb- the unit test , and I make migration for  "RelatedManager" ( MultipleChooserPanel) object  manually because when I tried to automatically run migrations some error occurs , so I had to make it manually ,so if you know how to make it automatically  without any error, please let me know and I will update this PR , also I just added one unit test so if you have other suggestions about unit tests please let me know.

elhussienalmasri commented:
when I run migration before I add any code this error occurs `  editor = editor_settings[name] KeyError: 'custom'` , and after 
this error occurred:
![Screenshot](https://github.com/wagtail/wagtail/assets/90080237/47dd5fd8-c603-4578-88be-c9a3b0bd2141), 
may it because I added two " MultipleChooserPanel" but I want to run migration automatically and solve the first error.
I just said that to take into consideration when review this PR, by the way I only want t add " MultipleChooserPanel" to 
exclude object " PageWithExcludedCopyField" , may if I can run migration automatically , may I find solution to this issue,
so any idea how to run migration automatically.

gasman commented on wagtail/test/testapp/migrations/0015_gallerypage_gallerypageimage.py:
```diff
@@ -66,6 +66,15 @@ class Migration(migrations.Migration):
                 ),
+                (
+                    "exclude_field",
+                    modelcluster.fields.ParentalKey(
```
Two ParentalKeys on the same model isn't really valid, because ParentalKey means "this model will only ever be edited as a child of the linked model". The test here isn't creating any GalleryPageImage instances (which it really should be doing if it's going to test that `exclude_fields_in_copy` is doing the right thing, rather than just confirming that it doesn't crash), but if it did, we'd have problems with the `page` key being non-null. It would be better to set up a dedicated child model for `PageWithExcludedCopyField` here.

gasman commented:
This is actually a more general problem - fields in the revision data that are listed in `exclude_fields_in_copy` are being replaced with fields from the live page object, without considering whether the field value is JSON-serialisable. This affects related fields (which give the Manager instance), but also StreamField (which returns a StreamValue, which is also non-JSON-serialisable). A better fix is to call `page_copy.serializable_data` up front, which returns the page data that would be written if we were saving a new revision, and use this as the source of data - I'll open a new PR using this approach shortly.
